crates/ty_python_semantic/src/types/call/arguments.rs

```diff
@@ -108,6 +117,72 @@ impl<'a, 'db> CallArgumentTypes<'a, 'db> {
     pub(crate) fn iter(&self) -> impl Iterator<Item = (Argument<'a>, Type<'db>)> + '_ {
         self.arguments.iter().zip(self.types.iter().copied())
     }
+
+    /// Returns an iterator on performing [argument type expansion].
+    ///
+    /// Each element of the iterator represents a set of argument lists, where each argument list
+    /// contains the same arguments, but with one or more of the argument types expanded.
+    ///
+    /// [argument type expansion]: https://typing.python.org/en/latest/spec/overload.html#argument-type-expansion
+    pub(crate) fn expand(&self, db: &'db dyn Db) -> impl Iterator<Item = Vec<Vec<Type<'db>>>> + '_ {
+        /// Represents the state of the expansion process.
+        ///
+        /// This is useful to avoid cloning the initial types vector if none of the types can be
+        /// expanded.
+        enum State<'a, 'db> {
+            Initial(&'a Vec<Type<'db>>),
+            Expanded(Vec<Vec<Type<'db>>>),
+        }
```

Comment:
This mainly an optimization to avoid cloning the `self.types` vector to initialize the `std::iter::successors` iterator. This would be useful in cases where no argument types can be expanded.